packages/cli/fern-definition/validator/src/rules/compatible-ir-versions/compatible-ir-versions.ts

```diff
@@ -0,0 +1,95 @@
+import { createFdrGeneratorsSdkService } from "@fern-api/core";
+import { isVersionAhead } from "@fern-api/semver-utils";
+
+import { Rule, RuleViolation } from "../../Rule";
+
+function getMaybeBadVersionMessage(
+    generatorName: string,
+    minCliVersion: string,
+    cliVersion: string
+): RuleViolation[] | undefined {
+    if (!isVersionAhead(cliVersion, minCliVersion)) {
+        return [
+            {
+                severity: "error",
+                message: `The generator ${generatorName} requires CLI version ${minCliVersion} or later (current version: ${cliVersion}).`
+            }
+        ];
+    }
+    return;
+}
+
+export const CompatibleIrVersionsRule: Rule = {
+    name: "compatible-ir-version",
+    create: async () => {
+        return {
+            generatorsYml: {
+                generatorInvocation: async (invocation) => {
+                    const fdr = createFdrGeneratorsSdkService({ token: undefined });
+                    // TODO: is there a better way to do this? I feel like I can't bring a dep on the cli package
+                    // I also feel like we shouldn't throw or error if we can't get the CLI version, so just returning here.
+                    const cliVersion = process.env.CLI_VERSION;
+                    if (cliVersion == null) {
+                        return [];
+                    }
+
+                    // Pull the CLI release to get the IR version
+                    const cliRelease = await fdr.generators.cli.getCliRelease(cliVersion);
+                    if (!cliRelease.ok) {
+                        return [];
+                    }
+                    const cliIrVersion = cliRelease.body.irVersion;
+
+                    // Pull the generator release to get the IR version
+                    let invocationIrVersion: number;
+                    if (invocation["ir-version"] != null) {
+                        // You've overridden the IR version in the generator invocation, let's clean it up
+                        invocationIrVersion = Number(invocation["ir-version"].replace("v", ""));
+                        if (cliIrVersion >= invocationIrVersion) {
+                            // You've overridden the IR version and you're on a CLI that works with it!
+                            return [];
+                        }
+                    } else {
+                        const generatorEntity = await fdr.generators.getGeneratorByImage({
+                            dockerImage: invocation.name
+                        });
+                        if (!generatorEntity.ok) {
+                            return [];
+                        }
+                        const generatorRelease = await fdr.generators.versions.getGeneratorRelease(
+                            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+                            generatorEntity.body!.id,
```

Comment:
why the `!` operator? worth it to do a null check to avoid a runtime error

Reply:
For some reason the check above for `ok` does not force this to be non-null, but this should be a safe assertion

Reply:
It doesnt because the response body is optional [link](https://github.com/fern-api/fern-platform/blob/2f8d3341623e7d4428a6176279a4b11d7d06bfe0/fern/apis/fdr/definition/generators/__package__.yml#L71-L76)

Feel pretty strongly about never using `!` in our code